File: Offer 26. 树的子结构.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSubStructure(TreeNode A, TreeNode B) {
        if (A == null || B == null) {
            return false;
        } 
        return search(A, B) || isSubStructure(A.left, B) || isSubStructure(A.right, B);
    }

    boolean search(TreeNode A, TreeNode B) {
        // 这里即使B为空，B也是至少有一个节点，因此返回true
        if (B == null) {
            return true;
        }

        if (A == null || A.val != B.val) {
            return false;
        }

        return search(A.left, B.left) && search(A.right, B.right);
    }
}

File: Offer 26. 树的子结构.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSubStructure(self, A: TreeNode, B: TreeNode) -> bool:
        def recur(A, B):
            if not B:
                return True
            if not A:
                return False
            return A.val == B.val and recur(A.left, B.left) and recur(A.right, B.right)
        
        if not A or not B:
            return False
        return recur(A, B) or self.isSubStructure(A.left, B) or self.isSubStructure(A.right, B)